contracts/FreeMint.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: MIT
+// Compatible with OpenZeppelin Contracts ^5.0.0
+pragma solidity ^0.8.22;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {ERC721Pausable} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721Pausable.sol";
+import {ERC721URIStorage} from "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+/// @title FreeMint NFT Contract
+/// @notice Contract for minting NFTs with free mint capability
+contract FreeMint is ERC721, ERC721URIStorage, ERC721Pausable, Ownable {
+    using Strings for uint256;
+
+    uint256 private _nextTokenId;
+
+    string public baseContractURI;
+
+    string public baseURI;
+
+    string public baseExtension;
+
+    uint256 public maxSupply;
+
+    uint256 public mintLimit;
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _baseURI,
+        string memory _baseExtension,
+        uint256 _maxSupply,
+        uint256 _mintLimit,
+        uint256 _startWithTokenId
+    ) ERC721(_name, _symbol) Ownable(msg.sender) {
+        setBaseURI(_baseURI);
+        baseExtension = _baseExtension;
+        maxSupply = _maxSupply;
+        mintLimit = _mintLimit;
+        require(_startWithTokenId <= 1, "Invalid startWithTokenId");
+        _nextTokenId = _startWithTokenId;
+    }
+
+    function mint(address to, uint256 amount) public whenNotPaused {
+        require(amount > 0, "Invalid amount");
+        require(_nextTokenId + amount < maxSupply, "Max supply reached");
```

Comment:
Should be `<=`
Or it will not mint the last tokenId

Reply:
thanks, it is even more complex. Some projects start from tokenId=0. that also needs to be fixed.

Reply:
fixed and added many tests for it